### PR TITLE
pulley: superinstructions for pushing/popping list of registers

### PR DESCRIPTION
**Pull request by Kmeakin**
Adds superinstructions for pushing/popping any subset of XRegs in one go

### CODE REVIEW

**Comment by fitzgen** [comment]
And then once https://github.com/bytecodealliance/wasmtime/pull/9088 is rebased and lands, we can do the same with this one. Thanks again for making these improvements @Kmeakin!

**Comment by fitzgen** [comment]
This needs a rebase after #9088

File: pulley/fuzz/src/interp.rs
Comment (fitzgen) [inline pulley/fuzz/src/interp.rs:116]: These are only safe if we have enough stack space to push to or pop from, right?

So either we should make these not safe for fuzzing, or we should keep track of how much stack we've consumed and still have available at the point where this `op` is executed and gate `xpush`/`xpop`/`push_frame`/`pop_frame`'s fuzzing safety on whether there is enough stack to safely execute them.

**Review by fitzgen — COMMENTED** [review]
Looks great! Just a couple things to address below before we merge this.

File: cranelift/bitset/src/compound.rs
Comment (Kmeakin) [reply cranelift/bitset/src/compound.rs:349]: `ScalarBitset` now implements `Ord` (because pulley `Instr` derives `Ord`). So calls to `bitset.min()` will resolve to the `min` method from `Ord`, rather than the inherent method attached to `ScalarBitset`

File: pulley/src/interp.rs
Comment (alexcrichton) [inline pulley/src/interp.rs:1071]: Are there perhaps some typos here? Otherwise `sp` is set twice. I also wouldn't expect `sp` to be set here at all other than through the side-effects of `pop()`?

File: cranelift/bitset/src/compound.rs
Comment (fitzgen) [reply cranelift/bitset/src/compound.rs:349]: Can we avoid `Ord` for `ScalarBitSet`? And instead implement it for `RegSet`?

File: pulley/fuzz/src/interp.rs
Comment (Kmeakin) [reply pulley/fuzz/src/interp.rs:116]: I'll return `false` for now, we can add a static analysis later

File: cranelift/bitset/src/compound.rs
Comment (Kmeakin) [reply cranelift/bitset/src/compound.rs:349]: We can remove the `#[derive(PartialOrd, Ord)]` on `Op`, since it's not used anywhere.

File: pulley/src/interp.rs
Comment (Kmeakin) [reply pulley/src/interp.rs:1071]: I didn't reverse the logic from `push_frame` completely. IIUC, sp should be restored from fp before popping fp and lr.

**Review by fitzgen — APPROVED** [review]
Looks great, thanks!

I hope to get the runtime integration soon, and then we can start enabling Wasm spec tests and such one by one until we have everything running under Pulley!

**Comment by alexcrichton** [comment]
I believe the error in CI is that in the `scripts/publish.rs` file the `cranelift-bitset` crate needs to be reordered before the `pulley` family of crates since they now depend on `cranelift-bitset`

**Comment by fitzgen** [comment]
Side note, it will be a little faster to iterate here to add an empty commit with the message `prtest:full` so that the full CI checks happen and we don't have to wait for the merge queue.

**Comment by fitzgen** [comment]
I'm not totally sure what is going on here, new-ish dependencies for new-ish are always a little annoying to get set up correctly so that when we publish we don't run into any issues. I guess we are effectively front-loading the issues, rather than dealing with them at publish time, which is probably the right choice. Anyways, I think you should be able to repro locally with

```shell-session
$ rustc scripts/publish.rs
$ ./publish verify
```

Note that if you have a `.cargo` directory in your wasmtime checkout already, you'll want to move it somewhere else temporarily while running the above.

**Comment by fitzgen** [comment]
> I believe the error in CI is that in the `scripts/publish.rs` file the `cranelift-bitset` crate needs to be reordered before the `pulley` family of crates since they now depend on `cranelift-bitset`

^ Did the fix for this perhaps get lost in a rebase?

**Pull request merged**
